[New] `install`: add `-b` flag to skip building from source

sladyn98 opened the pull request:
This PR aims to add the `--b` flag that skips source download and  install from binary only.'
Closes : https://github.com/nvm-sh/nvm/issues/2114

sladyn98 commented on nvm.sh:
```diff
@@ -1935,6 +1935,12 @@ nvm_install_binary() {
 
+  local nosource
+  # Read nosource from arguments
+  if [ $nosource -eq 1]; then
```
@ljharb What would be the best way to read the `-b` flag here. I could not see the `-s` flag being read anywhere. I just put a boolean here and thought that we could read the flag and set it accordingly

ljharb replied:
the `-s` flag is read here: https://github.com/nvm-sh/nvm/blob/master/nvm.sh#L2901-L2903

and you'd need to pass it into `nvm_install_binary` here: https://github.com/nvm-sh/nvm/blob/master/nvm.sh#L3168 as either an arg, or an env var.

ljharb commented on nvm.sh:
```diff
@@ -2665,6 +2671,7 @@ nvm() {
         nvm_echo '  nvm install [<version>]                     Download and install a <version>. Uses .nvmrc if available and version is omitted.'
         nvm_echo '   The following optional arguments, if provided, must appear directly after `nvm install`:'
         nvm_echo '    -s                                        Skip binary download, install from source only.'
+        nvm_echo '    -b                                        Skip source download, install from binary only.'
```
we should ensure that `-s -b` errors.

sladyn98 replied:
Yeah so with respect to this if the user specifies `-s` line number https://github.com/nvm-sh/nvm/blob/master/nvm.sh#L3167 will not execute anyways so I was thinking maybe pass the {nosource} parameter to `nvm_install_source` and then check if its set inside we throw an error ? Does this approach sound okay ?

ljharb replied:
I think it's better to check for conflicting arguments at argument parse time, so that those errors come as soon as possible - with your suggestion, if there's no binary available on nodejs.org, it wouldn't error on `-s -b`.

sladyn98 replied:
Yeah made the change

ljharb commented on nvm.sh:
```diff
@@ -2911,10 +2911,18 @@ nvm() {
           -s)
             shift # consume "-s"
             nobinary=1
+            if ["${nosource}" -eq 1]; then
```
```suggestion
            if [ $nosource -eq 1 ]; then
```
quotes aren't needed for -eq

ljharb commented on nvm.sh:
```diff
@@ -2911,10 +2911,18 @@ nvm() {
           -b)
             shift #consume "-b"
             nosource=1
+            if ["${nobinary}" -eq 1]; then
```
```suggestion
            if [ $nobinary -eq 1 ]; then
```

ljharb commented on nvm.sh:
```diff
@@ -1926,6 +1926,9 @@ nvm_install_binary() {
   fi
 
+  local nosource
+  nosource = "${4-}"
```
```suggestion
  nosource="${4-}"
```
assignments won't work if there's spaces around the =

ljharb commented on nvm.sh:
```diff
@@ -1965,6 +1968,12 @@ nvm_install_binary() {
 
+
+  # Read nosource from arguments
+  if [ $nosource -eq 1]; then
+      nvm_err "Binary download failed"
+      return 0
```
```suggestion
  if [ $nosource -eq 1 ]; then
      nvm_err "Binary download failed"
      return 0
  fi
```

ljharb commented on nvm.sh:
```diff
@@ -2901,7 +2911,18 @@ nvm() {
           -s)
             shift # consume "-s"
             nobinary=1
```
see above; you'll need to declare `nosource` the same way `nobinary` is declared.

sladyn98 commented:
I am guessing we would need to write some tests for this as well :)

ljharb reviewed (commented):
Looks great! Now we just need some tests :-)

 - unit test for passing `-b`, showing that it passes through to `nvm_install_binary` (i'm assuming there's already a similar one for `-s`; if not, let's add one!)
  - unit test for `nvm_install_binary` that shows that when `nosource` is passed, it does not attempt to install from source
   - (unit or "installation") test that would normally fall back to installing from source, showing that the error happens when nosource is set

ljharb commented on test/fast/Unit tests/nvm_install_binary_nosource:
```diff
@@ -0,0 +1,14 @@
+    echo "NOSOURCE:${nosource-}
+}
+
+OUTPUT="$(nvm install -b 8.0.0)"
+EXPECTED_OUTPUT='NOSOURCE:1'
+[ "${OUTPUT}" = "${EXPECTED_OUTPUT}" ] || die "No source binary flag is active and should have returned true. Instead it returned >${OUTPUT}<"
```
```suggestion
OUTPUT="$(nvm install -b 8.0.0)"
EXPECTED_OUTPUT='NOSOURCE:1'
[ "${OUTPUT}" = "${EXPECTED_OUTPUT}" ] || die "No source binary flag is active and should have returned true. Instead it returned >${OUTPUT}<"

OUTPUT="$(nvm install 8.0.0)"
EXPECTED_OUTPUT='NOSOURCE:0'
[ "${OUTPUT}" = "${EXPECTED_OUTPUT}" ] || die "No source binary flag is not active and should have returned false. Instead it returned >${OUTPUT}<"
```
this way we've covered both angles :-)

sladyn98 commented on test/fast/Unit tests/nvm_install_binary_nosource:
```diff
@@ -0,0 +1,32 @@
+# Check if nosource stops downloading from source
+
+set +ex # needed for stder
+OUTPUT="$(nvm install -b 8.0.0 2>&1)"
```
@ljharb Just not sure how would we make the binary download fail here. Maybe pass a version that did not have a binary version available

ljharb replied:
we could override `nvm_binary_available` to return 1

sladyn98 replied:
Yeah it looks cool now

sladyn98 commented:
@ljharb The tests look good now. Do let me know if there are any more we need to add ?

ljharb commented on test/installation_node/install from binary:
```diff
@@ -12,7 +12,7 @@ NVM_TEST_VERSION=v0.10.7
 
 # Install from binary
-nvm install $NVM_TEST_VERSION || die "install $NVM_TEST_VERSION failed"
+nvm install -b $NVM_TEST_VERSION || die "install $NVM_TEST_VERSION failed"
```
it'd probably be good to duplicate this test file, so we have tests for this both with and without the `-b` flag?

sladyn98 replied:
Got it :+1:

ljharb reviewed (commented):
it'd still be good to have a (unit or "installation") test that would normally fall back to installing from source, showing that the error happens when nosource is set. otherwise, LGTM!

sladyn98 commented:
> it'd still be good to have a (unit or "installation") test that would normally fall back to installing from source, showing that the error happens when nosource is set. otherwise, LGTM!

Isn't this test already there ?  we are already passing the -b flag which sets nosource and then it displays the error that skipping download from source

ljharb commented:
hm, the build seems to be timing out. is there a chance the "fast" tests are trying to do an actual installation?

sladyn98 commented:
> hm, the build seems to be timing out. is there a chance the "fast" tests are trying to do an actual installation?

Yeah I moved the installation test back to the installation directory and modified the fast tests to run correctly.

ljharb commented on test/fast/Unit tests/nvm_install_binary_nosource:
```diff
@@ -0,0 +1,32 @@
+# Unit test to check if the function errors out when the flag is set
+OUTPUT="$(nvm_install_binary node std 8.0.0 1 2>&1)"
+EXPECTED_OUTPUT='Binary download failed.Download from source aborted'
+if [[ "${OUTPUT}" != *"${EXPECTED_OUTPUT}"* ]]; then
```
these are bash-only comparisons which i'd expect to fail in sh/dash/ksh.

i'll update them to use posix comparisons.

sladyn98 replied:
Sure

ljharb commented on nvm.sh:
```diff
@@ -1965,6 +1968,13 @@ nvm_install_binary() {
+
+  # Read nosource from arguments
+  if [ "$nosource" -eq 1 ]; then
+      nvm_err "Binary download failed.Download from source aborted"
```
```suggestion
      nvm_err 'Binary download failed. Download from source aborted.'
```

ljharb commented on nvm.sh:
```diff
@@ -1965,6 +1968,13 @@ nvm_install_binary() {
 
+
+  # Read nosource from arguments
+  if [ "$nosource" -eq 1 ]; then
```
```suggestion
  if [ "${nosource-}" = '1' ]; then
```
i think this one's a string at this point
